game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -271,80 +272,110 @@ public Territory getTerritory(final String s) {
   }
 
   /**
-   * Returns the shortest route between two territories or null if no route exists.
+   * Returns all neighbors within a certain distance of the starting territory that match the
+   * condition. Does NOT include the original/starting territory in the returned Set.
    *
-   * @param t1 start territory of the route
-   * @param t2 end territory of the route
+   * <p>TODO: update to properly consider movement cost not just distance
+   *
+   * @param territory referring territory
+   * @param movementLeft maximal distance of the neighboring territories
    */
-  public Route getRoute(final Territory t1, final Territory t2) {
-    return getRoute(t1, t2, Matches.territoryIsLandOrWater());
+  public Set<Territory> getNeighborsByMovementCost(
+      final Territory territory, final Unit unit, final BigDecimal movementLeft) {
+    Preconditions.checkArgument(
```

Comment:
Good to see the precondition check, given that `Route` can have null territories, I wonder if it would be more robust include another precondition check for non-null territory and non-null unit.

Reply:
Added a check for unit. Territory is actually already checked in getNeighbors.

Reply:
Public APIs tend to be treated a bit special as they can be invoked from new locations and one cannot generally assume that you know all the caller paths into the method. Private methods are a bit of an exception because if you raise the scope of a method, then it is a buyer-beware situation.

It's one benefit even to reducing scope, you can be less stringent about validation and asserting state. If we don't get a NPE in very short order, 'fail-early' principle says to fail as soon as an error condition could be detected, which is either the caller of this method or as soon as the method starts executing.

I guess in other ways, it doesn't hurt, does it?

Reply:
I mean I think you can go too far with Preconditions and add them to every method for every argument. Its just a lot of boiler plate code that really if you are already checking it in another place is just doing nothing. There is definitely value in many cases of failing fast and ensuring things like NPEs don't happen deep inside methods. But there are also logical boundaries to how things are used. There are a lot of similar methods in this call due to offering different parameter combinations that then use a private method to check the territory isn't null.

Reply:
Hmm, I still think that bakes in assumptions about how the public method is used. The limit is validity of all arguments passed in. NPE checks are sometimes not needed as you'll get a NullPointer in the first few lines anyways. If a value is unexpectedly null and then is propagated deeper into the system, it's a failure to detect an error condition. I don't think adding one extra line of check is going to really hurt or make this harder to understand.